### ISSUE TITLE
handshake simulation doesn't take curves into account

### ISSUE THREAD

**Issue by drwetter** [drwetter/testssl.sh#375]
Currently testssl.sh uses ``secp521r1``. For some windows clients (according to SSLlabs: "IE 10 / Win Phone 8.0" "IE 11 / Win Phone 8.1 "  this EC isn't supported.

So e.g. scanning testssl.sh SSLlabs gives currently a handshake failure whereas ``testssl.sh -c testssl.sh`` yields:
```
IE10 / Win Phone 8.0          TLSv1.0 ECDHE-RSA-AES256-SHA
IE11 / Win Phone 8.1          TLSv1.2 ECDHE-RSA-AES256-SHA
```

**Comment by drwetter** [comment]
Hi @seccubus,

any idea how to tackle this?

Cheers, Dirk

**Comment by seccubus** [comment]
I need to figure out a way to find out which curves are supported by which browsers. Then it should be possible.

**Comment by drwetter** [comment]
Hi @seccubus,

had to search for it -- as I was sure remembering that. Here you go: https://api.dev.ssllabs.com/api/v3/getClients, taken from https://github.com/ssllabs/ssllabs-scan/issues/295#issuecomment-172566653 .

The curves are in ``etc/curves.txt``.

Cheers, Dirk

**Comment by seccubus** [comment]
Thanks, let me have a think about this.

**Comment by seccubus** [comment]
I want to figure out a way to easily integrate Ivan 's work here.

**Comment by drwetter** [comment]
Take your time!

A solid solution probably means that the server side curves will be identified properly. This hasn't been done yet #123. 

Besides this week I am really swamped with paid work .

**Comment by dcooper16** [comment]
I noticed that https://api.dev.ssllabs.com/api/v3/getClients includes the field `hexHandshakeBytes` for each client. So, rather than just extracting information such as the cipher suites and curves supported from each client and then using OpenSSL, the `hexHandshakeBytes` could be used with something like `tls_sockets()` to even more closely simulate the behaviour of each client. `tls_sockets()` can't handle starttls, and I notice that the call to `$OPENSSL s_client` in `run_client_simulation()` includes `$STARTTLS`, but `run_client_simulation()` only executes if `$SERVICE` is HTTP, so I don't think `$STARTTLS` applies.

It would be somewhat non-trivial to use the `hexHandshakeBytes` fields. I parsed one of them and saw that it included a server name extension with a value of "dev.ssllabs.com," but it would be possible to parse the `hexHandshakeBytes` field and either remove the server name extension (if $SNI is empty) or replace it with the appropriate value for the server (if $SNI is not empty). It would take some time, but it is something I could try to do, if you think that would be the best long-term solution.

**Comment by seccubus** [comment]
One of the benefits of testssl.sh over ssllabs is to use it for non https stuff. I'd prefer not to lose that ability

**Comment by seccubus** [comment]
I would also like to add some non-http clients like mail clients.

**Comment by dcooper16** [comment]
testssl.sh may be used for non-HTTPS stuff, but `run_client_simulation()` can't:
```
run_client_simulation() {
     # Runs browser simulations. Browser capabilities gathered from:
     # https://www.ssllabs.com/ssltest/clients.html on 10 jan 2016
     local names=()
     local short=()
     local protos=()
     local ciphers=()
     local tlsvers=()
     local sni=()
     local warning=()
     local i=0

     # doesn't make sense for other services
     if [[ $SERVICE != "HTTP" ]];  then
          return 0
     fi
     ...
}
```
My proposal wouldn't preclude adding other client types, using OpenSSL could be used for those client types that require starttls. But `run_client_simulation()` currently just tests HTTPS clients, and it seems that those clients could be more accurately simulated by copying the actual handshake messages rather than just sending the closest approximation of the handshake messages that can be created using OpenSSL.

**Comment by seccubus** [comment]
@drwetter added that if statement, it wasn't what I had in mind.

**Comment by seccubus** [comment]
I understand your point btw. But that would means two branches in the code one for startssl the other for starttls (then again browsers down support it)

**Comment by drwetter** [comment]
> I noticed that https://api.dev.ssllabs.com/api/v3/getClients includes the field hexHandshakeBytes 
> for each client. So, rather than just extracting information such as the cipher suites and curves 
> supported from each client and then using OpenSSL, the hexHandshakeBytes could be used 
> with  something like tls_sockets() to even more closely simulate the behaviour of each client.

good idea. 

> tls_sockets() can't handle starttls,

nope :-)   The neat thing is that fd_socket() does the STARTTLS handshake or the CONNECT to the proxy. Then testssl.sh is agnostic / it doesn't care.

**Comment by drwetter** [comment]
> @drwetter added that if statement, it wasn't what I had in mind. 

There was a reason I can't recall after my long day (or only recall vaguely) but it's somewhere in github..

**Comment by drwetter** [comment]
@dcooper16 ``testssl.sh -t  smtp -p  -D aspmx.l.google.com.:25`` (SSLv3-TLS 1.2 needs to be tested, then it should be ok to use sockets here, too)

**Comment by seccubus** [comment]
At that time there were no non-http clients in the list if I recall correctly.

**Comment by dcooper16** [comment]
Parsing the hexHandshakeBytes from https://api.dev.ssllabs.com/api/v3/getClients and then regenerating them with the correct server name extension values for the server being tested turned out to be easier than I thought.

I posted the code that I developed as a PR (#386), but I don't expect it to be merged. I just posted it to give people a chance to test it out.

As written the $EXPERIMENTAL flag needs to be set; otherwise OpenSSL will be used.

I included an implementation of rfc2openssl() in the PR just so that the output would look the same whether OpenSSL or sockets is being used. An alternative would have been to print the RFC names for the ciphers when sockets are being used.

During my testing I encountered some instances in which the OpenSSL and sockets implementations created different results where I felt that the information for the OpenSSL implementation was incorrect (e.g., ciphers, protos, or tlsvers). Where that happened I corrected the OpenSSL-related information in this PR. I could submit those changes as a separate PR, but not until late next week.

I don't understand the relevance of `testssl.sh -t smtp -p -D aspmx.l.google.com.:25`, but I'm limited in my ability to test that one, since I can't test it when I'm at work (the firewall blocks access to port 25).

I also added the sockets version of run_client_simulation() to https://github.com/dcooper16/testssl.sh/blob/extended_tls_sockets/testssl.sh. In that version, for each cipher that uses an ephemeral key, read_dhbits_from_file() is used to display the algorithm (DH or ECDH) and number of bits.

**Comment by seccubus** [comment]
@dcooper16 please have a look at https://github.com/dcooper16/testssl.sh/pull/1/files 

I've tried to make the client signatures more maintainable by splitting them into a separte file and having them generated by a program that takes its input from the ssllabs API.

**Comment by drwetter** [comment]
> I don't understand the relevance of testssl.sh -t smtp -p -D aspmx.l.google.com.:25, but I'm 
> limited  in my ability to test that one, since I can't test it when I'm at work (the firewall blocks 
> access to port 25).

You can also add -D (for DROWN). I was trying to explain that sockets are also used for STARTTLS.

**Comment by seccubus** [comment]
I can confirm fd_sockets work with starttls:

```
$ ./testssl.sh -t smtp -c smtp.sectionzero.org:587

###########################################################
    testssl.sh       2.7dev from https://testssl.sh/dev/
    (4aba0b9 2016-06-27 08:40:38 -- 1.507)

      This program is free software. Distribution and
             modification under GPLv2 permitted.
      USAGE w/o ANY WARRANTY. USE IT AT YOUR OWN RISK!

       Please file bugs @ https://testssl.sh/bugs/

###########################################################

 Using "OpenSSL 1.0.2-chacha (1.0.2e-dev)" [~181 ciphers]
 on sbpltc2ksg3qr:./bin/openssl.Darwin.x86_64
 (built: "Aug 27 20:29:14 2015", platform: "darwin64-x86_64-cc")


 Start 2016-06-27 08:54:02    -->> 178.237.34.227:587 (smtp.sectionzero.org) <<--

 rDNS (178.237.34.227):  a4091.mcehosting.atom86.net.
 Service set:            STARTTLS via SMTP

 Running browser simulations (experimental)

 Android 2.3.7                    TLSv1.0 AES128-SHA
 Android 4.0.4                    TLSv1.0 DHE-RSA-AES256-SHA
```

I will modify @dcooper16 's code so we will actually do handshake simulations for certain clients if the protocol isn't HTTP. For certain situations (e.g. ldap we will still need native openssl support).

**Comment by seccubus** [comment]
This can be closed if PR #394 is accepted and the $EXPERIMENTAL if statement is removed from run_client_sims()

**Comment by drwetter** [comment]
Why does this depend on #394? 

The open point here is some testing. Right now I see  results at least on https://testssl.sh which are good 

![diff](https://cloud.githubusercontent.com/assets/8036727/16444241/6cbfb1ce-3ddd-11e6-8218-7cce2ff09481.png)

and match SSLlabs (see right hand side): https://www.ssllabs.com/ssltest/analyze.html?d=testssl.sh

**Comment by seccubus** [comment]
Currently testing via openssl is broken for for e.g. services that offer only tls1.2. 
Also #394 enables socket based testing for starttls services and other non-http stuff. 
If we implement those changes socket based testing can be the default and it does take curves into account.

**Comment by drwetter** [comment]
Am 29. Juni 2016 11:23:43 MESZ, schrieb Frank Breedijk <notifications@github.com>:
>Currently testing via openssl is broken for for e.g. services that
>offer only tls1.2. 
>Also #394 enables socket based testing for starttls services and other
>non-http stuff. 
>If we implement those changes socket based testing can be the default
>and it does take curves into account. 

Yes! I was interrupted while writing the mail. 

Probably testing AND enabling sockets should do it, right?

Cheers, Dirk



-- 
Set from my mobile. Excuse my brevity&typos

**Comment by seccubus** [comment]
If updated PR #394 to include pull requests. Found out it isn't perfect yet.
